LFS-1069: Single select upload function should not allow multiple files selection: fixed regression in FileResourceQuestion with maxAnswers > 1

marta- opened the pull request.

marta- committed:
[misc] Fixed error display for FileResourceQuestion

marta- committed:
LFS-1069: Single select upload function should not allow multiple files selection
Fixed regression in FileResourceQuestion with maxAnswers > 1

sashaandjic reviewed (approved):
tested

sdumitriu reviewed (commented):
I can't find a simple way to reproduce it consistently, but sometimes I am able to upload 3 files when maxAnswers=2. I'll keep on trying to find a pattern.

marta- committed:
LFS-1069: Single select upload function should not allow multiple files selection
Fixed bug that allowed to upload more than maxAnswers files for maxAnswers > 1

sashaandjic commented:
**Test case for MaxAnswers = 1**
![image](https://user-images.githubusercontent.com/23243244/114348081-0f2acd00-9b34-11eb-88cf-b2d2ec10e522.png)

**I tried to upload 3 files:**
![image](https://user-images.githubusercontent.com/23243244/114348291-5618c280-9b34-11eb-8db3-932d85edc5d5.png)

It appeared only one was available for the upload **123_brain_slice3.csv**:
![image](https://user-images.githubusercontent.com/23243244/114348359-6af55600-9b34-11eb-913e-07042a261359.png)

However, when I clicked on the **UPLOAD** button, another file got uploaded **123_brain_slice4** and one more file appeared to be ready for an upload **123-brain_slice1**:
![image](https://user-images.githubusercontent.com/23243244/114348786-01297c00-9b35-11eb-8d43-3cecee96d806.png)
 
After I clicked on the **UPLOAD** button again, the other 2 files were uploaded:
![image](https://user-images.githubusercontent.com/23243244/114348566-ba3b8680-9b34-11eb-983d-b663f8d8a44c.png)

and as expected, 3 tumor regions were created:
![image](https://user-images.githubusercontent.com/23243244/114348860-1d2d1d80-9b35-11eb-8118-c458e31f2a33.png)

marta- commented:
This PR doesn't touch the Variant Upload page, only File resource questions in forms. The Variant Upload will be fixed separately in LFS-535 and LFS-832.

marta- commented:
> I can't find a simple way to reproduce it consistently, but sometimes I am able to upload 3 files when maxAnswers=2. I'll keep on trying to find a pattern.

Steps:
* create 3 files, A,B,C
* change SomaticVariants to have maxAnswers=2
* create a SomaticVariants form
* Select files A,C to upload => A and C get uploaded
* Select A,B,C for upload => A,B and C get uploaded

marta- committed:
LFS-1069: Single select upload function should not allow multiple files selection
Fixed bug that allowed to upload more than maxAnswers files for maxAnswers > 1

sdumitriu reviewed (changes requested):
- edit Somatic Variants, set maxAnswers=2
- create a new Somatic Variants form
- upload f1 -> bug, the file isn't uploaded at all, even though it appears to be
- upload f1, f2, f3 -> bug, only f1 appears uploaded, the widget reports that at most 2 files can be uploaded, so f2 and f3 were skipped

sdumitriu reviewed (dismissed):
Another bug: create new External File form (rehab mode), upload a file: "error, at most 0 files can be uploaded, not uploaded: <empty, the file appears uploaded>"

sdumitriu commented:
Oh, and I forgot to check on the latest version, but I get a JS error about not being able to call `clearData()` on the list of files.

marta- commented:
> Another bug: create new External File form (rehab mode), upload a file: "error, at most 0 files can be uploaded, not uploaded: <empty, the file appears uploaded>"

I don't think you're testing the latest code

marta- commented:
> * edit Somatic Variants, set maxAnswers=2
> * create a new Somatic Variants form
> * upload f1 -> bug, the file isn't uploaded at all, even though it appears to be

I'm not following. WDYM it's not uploaded at all? For me it gets uploaded when you save the form

> * upload f1, f2, f3 -> bug, only f1 appears uploaded, the widget reports that at most 2 files can be uploaded, so f2 and f3 were skipped

sdumitriu commented:
> WDYM it's not uploaded at all?

Ah, another bug, the files I was testing with were empty, and empty files don't get uploaded... Let me test with some non-empty files.

sdumitriu commented:
Non-empty files are correctly uploaded.

"at most 0 files" has indeed been fixed while I was testing it.

Still happening: refusing to upload one extra file when 1 is already there, max is 2, and trying to upload 3 (including the one already uploaded). If the already uploaded one is not in the selection, one more file is correctly added.

Still happening: the JS error about `clearData()`, but only in Firefox (Chrome seems OK).

sdumitriu commented:
The JS error is correct, it signals incorrect usage of that method. According to [the MDN documentation](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/clearData):

> This method can only be used in the handler for the `dragstart` event, because that's the only time the drag operation's data store is writeable.

And also:

> This method does not remove files from the drag operation

marta- committed:
LFS-1069: Single select upload function should not allow multiple files selection
Fixed bug that allowed to upload more than maxAnswers files for maxAnswers > 1

sashaandjic reviewed (approved):
tested

sashaandjic commented:
Draging more than one file when maxAnswers = 1 will upload the file where the cursor was pointing. The error message is incomplete. I initially had C file uploaded, then dragged A, B and C with my mouse pointing on A file in the group. The file A got uploaded, and the error message did not specify that B and C were not uploaded.
![image](https://user-images.githubusercontent.com/23243244/114765421-2683d900-9d33-11eb-90cd-6ba00a88fb8f.png)

marta- committed:
LFS-1069: Single select upload function should not allow multiple files selection
Fixed bug with error display when dropping more than one file for maxAnswers = 1

sashaandjic reviewed (approved):
tested

OrdiNeu reviewed (approved):
Tested. Code looks good.
